### PR TITLE
[BEAM-5396] Savepoint restore option in Flink runner

### PR DESCRIPTION
**Pull request by tweise**
------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java [hunk 1]
```diff
@@ -204,4 +204,17 @@
   ExecutionMode getExecutionModeForBatch();
 
   void setExecutionModeForBatch(ExecutionMode executionMode);
+
+  @Description("Savepoint restore path.")
+  String getSavepointPath();
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Maybe extend the description a bit, also default set to `""`?

    If specified, attempts to restore the streaming pipeline from the provided Savepoint path.

Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Would be nice to test the default in `PipelineOptionstest#testdefaults`

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java [hunk 2]
```diff
@@ -204,4 +204,17 @@
...
+
+  void setSavepointPath(String path);
+
+  @Description(
+      "Flag indicating whether non restored state is allowed if the savepoint "
+          + "contains state for an operator that is no longer part of the pipeline.")
+  @Default.Boolean(false)
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Would be nice to test the default in `PipelineOptionstest#testdefaults`

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java [hunk 3]
```diff
@@ -138,11 +150,20 @@ static StreamExecutionEnvironment createStreamExecutionEnvironment(
     } else if (masterUrl.matches(".*:\\d*")) {
       List<String> parts = Splitter.on(':').splitToList(masterUrl);
       flinkConfig.setInteger(RestOptions.PORT, Integer.parseInt(parts.get(1)));
+
+      SavepointRestoreSettings savepointRestoreSettings = SavepointRestoreSettings.none();
+      if (options.getSavepointPath() != null) {
+        savepointRestoreSettings =
+            SavepointRestoreSettings.forPath(
+                options.getSavepointPath(), options.getAllowNonRestoredState());
+      }
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: Maybe it's just me but I prefer this pattern:

```java
      final SavepointRestoreSettings savepointRestoreSettings;
      if (options.getSavepointPath() != null) {
        savepointRestoreSettings =
            SavepointRestoreSettings.forPath(
                options.getSavepointPath(), options.getAllowNonRestoredState());
      } else {
        savepointRestoreSettings = SavepointRestoreSettings.none();
      }
```

It's slightly more verbose but also more robust.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java [hunk 4]
```diff
@@ -263,4 +284,81 @@ private static void applyLatencyTrackingInterval(
     long latencyTrackingInterval = options.getLatencyTrackingInterval();
     config.setLatencyTrackingInterval(latencyTrackingInterval);
   }
+
+  /**
+   * TODO: https://issues.apache.org/jira/browse/BEAM-5396
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: Can we move this after the class description?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java [hunk 5]
```diff
@@ -263,4 +284,81 @@ private static void applyLatencyTrackingInterval(
...
+
+      try {
+        return client
+            .run(streamGraph, jarFiles, globalClasspaths, usercodeClassLoader, restoreSettings)
+            .getJobExecutionResult();
+      } catch (ProgramInvocationException e) {
+        throw e;
+      } catch (Exception e) {
+        String term = e.getMessage() == null ? "." : (": " + e.getMessage());
+        throw new ProgramInvocationException("The program execution failed" + term, e);
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: Not sure about the `term` string. I think we can drop it.

**Comment by tweise** [comment]
Run Portable_Python PreCommit

**Comment by tweise** [comment]
I have tested this with the synthetic source example:
1. Run example: `python -m apache_beam.examples.flink.flink_streaming_impulse --parallelism=1 --flink_master=localhost:8081`
2. Cancel with savepoint: `./bin/flink cancel -s /Users/tweise/flink/flink-1.5.5/savepoints/1 2be8be7dc874ec2bd88c6c232fefeb45`
3. Restore: `python -m apache_beam.examples.flink.flink_streaming_impulse --parallelism=1 --flink_master=localhost:8081 --savepoint_path=file:/Users/tweise/flink/flink-1.5.5/savepoints/1/savepoint-2be8be-bbf6cce24652`

Then check the JM log to confirm the savepoint was used to resume.

Testing this only works with "remote execution" and involves the Flink CLI. Given there is coverage in Flink for this, not sure it is worth the complexity and effort?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java (hunk 5 above)
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: This is unchanged copied, intentionally. I want to keep it as is, and in the future replace with utility that lives in Flink.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java (hunk 3 above)
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: It's probably more a matter of taste, I made the change.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java (hunk 1 above)
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Done!

**Comment by mxm** [comment]
Very nice. Thanks!

**Pull request merged**